[Intelligence Effects] Add some methods to perform visual operations for a given range

rr-codes opened the pull request:
#### 098dc9dd7c74d4a378a429a1e3ef924339354fca
<pre>
[Intelligence Effects] Add some methods to perform visual operations for a given range
<a href="https://bugs.webkit.org/show_bug.cgi?id=281878">https://bugs.webkit.org/show_bug.cgi?id=281878</a>
<a href="https://rdar.apple.com/138351318">rdar://138351318</a>

Reviewed by Wenson Hsieh.

Expose some interfaces via the web content process to allow coordinating intelligence effects given a particular range:

* `proofreadingSessionSuggestionTextRectsInRootViewCoordinates`: Given a range, return the vector of rects that correspond
to the modified text post-replacement, limited to the text within the provided range.

* `updateTextVisibilityForActiveWritingToolsSession`: Hides or shows the text and document markers within the specified range.

* `textPreviewDataForActiveWritingToolsSession`: Takes a snapshot of the text in a particular range, temporarily forcing the
text to be visible.

These interfaces are encapsulated in a new protocol, `WKIntelligenceTextEffectCoordinatorDelegate`, which will be used by an
upcoming effects coordinator.

Subsequent changes will:

1. Factor out these interfaces from `WritingToolsController` to a new class that is agnostic of Writing Tools
2. Reduce code duplication for the preview generation.
3. Introduce `WKIntelligenceTextEffectCoordinator`.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebCore::Page::updateTextVisibilityForActiveWritingToolsSession):
(WebCore::Page::textPreviewDataForActiveWritingToolsSession):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebCore::WritingToolsController::updateTextVisibilityForActiveSession):
(WebCore::WritingToolsController::textPreviewDataForActiveSession):
(WebCore::WritingToolsController::currentState):
(WebCore::WritingToolsController::currentState const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView viewForIntelligenceTextEffectCoordinator:]):
(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:textPreviewsForRange:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebKit::WebPageProxy::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPageProxy::textPreviewDataForActiveWritingToolsSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebKit::WebPage::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPage::textPreviewDataForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_RectsForProofreadingSuggestionsInRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_TextPreviewsForRange)):

Canonical link: <a href="https://commits.webkit.org/285562@main">https://commits.webkit.org/285562@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9d101430bac02074e5a316633ea448ee19778edf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73023 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77231 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24261 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60257 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57399 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15886 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76090 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62853 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37817 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20323 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22590 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65909 "Found 1 new API test failure: TestWebKitAPI.WebKit.FindInUnifiedPDFAfterFindInPage (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78900 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65842 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-animation-fragmented.html imported/w3c/web-platform-tests/css/css-view-transitions/inline-element-size.html (failure)") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65119 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7105 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11243 "Built successfully and passed tests") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3039 "Built successfully") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h:
```diff
@@ -0,0 +1,52 @@
+- (NSView *)viewForIntelligenceTextEffectCoordinator:(id /* WKIntelligenceTextEffectCoordinator * */)coordinator;
+#endif
+
+- (void)intelligenceTextEffectCoordinator:(id /* WKIntelligenceTextEffectCoordinator * */)coordinator textPreviewsForRange:(NSRange)range completion:(void (^)(NSArray<id /* _WKTextPreview * */> *))completion;
```
fix for iOS

whsieh commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -929,6 +929,119 @@ static bool appliedCommandIsWritingToolsCommand(const Vector<Ref<WritingToolsCom
+    if (!textIndicator)
+        return std::nullopt;
+
+    // Add back any of the transparent content markers which were removed.
```
Nit - I think this comment could be removed if you renamed `markerRanges` to something like `transparentMarkerRangesToReinsert`

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2274,6 +2282,100 @@ - (void)writingToolsSession:(WTSession *)session didReceiveAction:(WTAction)acti
 
+#pragma mark - WKIntelligenceTextEffectCoordinatorDelegate conformance
+
+// FIXME: Give the `coordinator` parameters actual types.
```
Just curious — is there anything blocking this FIXME from being resolved? (Or is it going to be fixed in a subsequent patch)

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2274,6 +2282,100 @@ - (void)writingToolsSession:(WTSession *)session didReceiveAction:(WTAction)acti
+    return self;
+}
+
+- (void)intelligenceTextEffectCoordinator:(id /* WKIntelligenceTextEffectCoordinator * */)coordinator rectsForProofreadingSuggestionsInRange:(NSRange)range completion:(void (^)(NSArray<NSValue * /* CGRect/NSRect */> *))completion
```
Instead of `NSArray<NSValue * /* CGRect/NSRect */> *`, it might be clearer if you just made it `NSArray<NSValue *> *` (since the `rectsForProofreadingSuggestionsInRange:` part of the selector already implies that these are rect values). Alternately, you could consider using a `typedef` for the completion handler type.

whsieh commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -147,6 +147,7 @@
 		07297FA31C186ADB003F0735 /* WKUserMediaPermissionCheck.h in Headers */ = {isa = PBXBuildFile; fileRef = 07297FA11C186ADB003F0735 /* WKUserMediaPermissionCheck.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		074879B92373A90900F5678E /* AppKitSoftLink.h in Headers */ = {isa = PBXBuildFile; fileRef = 074879B72373A90900F5678E /* AppKitSoftLink.h */; };
 		074E75FE1DF2211900D318EC /* UserMediaProcessManager.h in Headers */ = {isa = PBXBuildFile; fileRef = 074E75FB1DF1FD1300D318EC /* UserMediaProcessManager.h */; };
+		074F34812CC5F7D500CA5301 /* WKIntelligenceTextEffectCoordinator.h in Headers */ = {isa = PBXBuildFile; fileRef = 074F34802CC5F7D500CA5301 /* WKIntelligenceTextEffectCoordinator.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Does this really need to be exposed as a new SPI header?

If so, it seems odd to omit it from the umbrella header (`WebKitPrivate.h`)

rr-codes replied:
it does, since it’s part of the WebKitSwift framework since the type will be implemented in Swift

whsieh replied:
Richard clarified offline that this is only marked private in the WebKitSwift library (not the WebKit framework)

rr-codes committed:
[Intelligence Effects] Add some methods to perform visual operations for a given range
https://bugs.webkit.org/show_bug.cgi?id=281878
rdar://138351318

Reviewed by Wenson Hsieh.

Expose some interfaces via the web content process to allow coordinating intelligence effects given a particular range:

* `proofreadingSessionSuggestionTextRectsInRootViewCoordinates`: Given a range, return the vector of rects that correspond
to the modified text post-replacement, limited to the text within the provided range.

* `updateTextVisibilityForActiveWritingToolsSession`: Hides or shows the text and document markers within the specified range.

* `textPreviewDataForActiveWritingToolsSession`: Takes a snapshot of the text in a particular range, temporarily forcing the
text to be visible.

These interfaces are encapsulated in a new protocol, `WKIntelligenceTextEffectCoordinatorDelegate`, which will be used by an
upcoming effects coordinator.

Subsequent changes will:

1. Factor out these interfaces from `WritingToolsController` to a new class that is agnostic of Writing Tools
2. Reduce code duplication for the preview generation.
3. Introduce `WKIntelligenceTextEffectCoordinator`.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebCore::Page::updateTextVisibilityForActiveWritingToolsSession):
(WebCore::Page::textPreviewDataForActiveWritingToolsSession):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebCore::WritingToolsController::updateTextVisibilityForActiveSession):
(WebCore::WritingToolsController::textPreviewDataForActiveSession):
(WebCore::WritingToolsController::currentState):
(WebCore::WritingToolsController::currentState const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView viewForIntelligenceTextEffectCoordinator:]):
(-[WKWebView intelligenceTextEffectCoordinator:rectsForProofreadingSuggestionsInRange:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:updateTextVisibilityForRange:visible:completion:]):
(-[WKWebView intelligenceTextEffectCoordinator:textPreviewsForRange:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebKit::WebPageProxy::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPageProxy::textPreviewDataForActiveWritingToolsSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceTextEffectCoordinator.h: Copied from Source/WebKit/WebKitSwift/WebKitSwift.h.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::proofreadingSessionSuggestionTextRectsInRootViewCoordinates const):
(WebKit::WebPage::updateTextVisibilityForActiveWritingToolsSession):
(WebKit::WebPage::textPreviewDataForActiveWritingToolsSession):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_RectsForProofreadingSuggestionsInRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_UpdateTextVisibilityForRange)):
(TEST(WritingTools, IntelligenceTextEffectCoordinatorDelegate_TextPreviewsForRange)):

Canonical link: https://commits.webkit.org/285562@main
